pkg/federate/kubefed/cluster.go

```diff
@@ -1,9 +1,299 @@
 package kubefed
 
 import (
-	"github.com/submariner-io/admiral/pkg/federate"
+	"fmt"
+	"reflect"
+	"time"
+
+	federate "github.com/submariner-io/admiral/pkg/federate"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	fedv1 "sigs.k8s.io/kubefed/pkg/apis/core/v1beta1"
+	"sigs.k8s.io/kubefed/pkg/client/generic/scheme"
+)
+
+const (
+	tokenKey = "token"
 )
 
 func (f *federator) WatchClusters(handler federate.ClusterEventHandler) error {
-	panic("not implemented")
+	f.startInformerOnce.Do(func() {
+		f.kubeFedClusterInformer.AddEventHandler(f)
+
+		go func() {
+			f.kubeFedClusterInformer.Run(f.stopChan)
+			klog.Info("KubeFedCluster informer stopped")
+		}()
+
+		klog.Info("Started KubeFedCluster informer")
+	})
+
+	var handlerString string
+	if stringer, ok := handler.(fmt.Stringer); ok {
+		handlerString = stringer.String()
+	} else {
+		handlerString = reflect.TypeOf(handler).String()
+	}
+
+	clusterWatcher := &clusterWatcher{
+		eventQueue:           make(chan func(), 1000),
+		enqueueTimerInterval: time.Duration(time.Second * 10),
+		stopChan:             f.stopChan,
+		handler:              handler,
+		handlerString:        handlerString,
+	}
+
+	f.Lock()
+	defer f.Unlock()
+
+	f.clusterWatchers = append(f.clusterWatchers, clusterWatcher)
+
+	go clusterWatcher.eventLoop()
+
+	for clusterID, kubeConfig := range f.clusterMap {
+		clusterWatcher.onAdd(clusterID, kubeConfig)
+	}
+
+	klog.Infof("Federator added ClusterEventHandler \"%s\"", handlerString)
+
+	return nil
+}
+
+func (f *federator) OnAdd(obj interface{}) {
+	klog.V(3).Infof("In federated cluster watcher OnAdd for %#v", obj)
+
+	kubeFedCluster := obj.(*fedv1.KubeFedCluster)
+
+	kubeConfig, err := f.buildFederatedClusterConfig(kubeFedCluster)
+	if err != nil {
+		klog.Errorf("Error building kube config for federated cluster %#v: %v", kubeFedCluster, err)
+		return
+	}
+
+	f.Lock()
+	defer f.Unlock()
+
+	if f.clusterMap[kubeFedCluster.Name] != nil {
```

Comment:
May be a reconnection to informer? I'm not sure though either, but if it happens I think it's a good thing to have it on the logs so we can understand something weird is going on.

Reply:
yeah that was my thinking too - being a bit paranoid 😄 If it did happen then at least I don''t  propagate a duplicate add event to our handlers.

Reply:
if you can do a test that triggers this scenario and then this code is needed, then having the code is fine. Otherwise, being paranoid doesn't help :)

Reply:
It's a simple check and doesn't hurt.